src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1312 @@
+        return prev_store.Save(prev_info);
+    }
+
+    CHIP_ERROR RemoveAllGroupMappings(chip::FabricIndex fabric_index, chip::EndpointId target_endpoint_id) override
+    {
+        VerifyOrReturnError(mInitialized, CHIP_ERROR_INTERNAL);
+        VerifyOrReturnError(kInvalidEndpointId != target_endpoint_id, CHIP_ERROR_INVALID_ARGUMENT);
+
+        FabricStore fabric_store(fabric_index);
+        FabricInfo fabric_info;
+
+        CHIP_ERROR err = fabric_store.Load(fabric_info);
+        VerifyOrReturnError(CHIP_NO_ERROR == err, CHIP_ERROR_INVALID_FABRIC_ID);
+        VerifyOrReturnError(kInvalidEndpointId != fabric_info.first_endpoint, CHIP_ERROR_KEY_NOT_FOUND);
+
+        // Existing fabric
+
+        chip::EndpointId endpoint_id   = fabric_info.first_endpoint;
+        chip::EndpointId prev_endpoint = kInvalidEndpointId;
+        EndpointInfo endpoint_info;
+        EndpointInfo prev_endpoint_info;
+        size_t endpoint_count = 0;
+
+        // Loop through the fabric's endpoints
+        do
+        {
+            EndpointStore endpoint_store(fabric_index, endpoint_id);
+            VerifyOrReturnError(CHIP_NO_ERROR == endpoint_store.Load(endpoint_info), CHIP_ERROR_KEY_NOT_FOUND);
+            if (endpoint_id == target_endpoint_id)
+            {
+                // Target endpoint info found
+                break;
+            }
+            prev_endpoint_info = endpoint_info;
+            prev_endpoint      = endpoint_id;
+            endpoint_id        = endpoint_info.next;
+        } while (++endpoint_count < fabric_info.endpoint_count);
+        VerifyOrReturnError(endpoint_id == target_endpoint_id, CHIP_ERROR_KEY_NOT_FOUND);
+
+        // Target endpoint found
+
+        chip::GroupId group_id   = endpoint_info.first_group;
+        chip::GroupId prev_group = kUndefinedGroupId;
+        GroupInfo group_info;
+
+        // Remove endpoint's groups
+        do
+        {
+            GroupMapStore group_store(fabric_index, endpoint_id, group_id);
+            ReturnErrorOnFailure(group_store.Load(group_info));
+            group_store.Delete();
+            prev_group = group_id;
+            group_id   = group_info.next;
+        } while ((kUndefinedGroupId != group_id) && (prev_group != group_id));
+
+        // endpoint_info.first_group = kUndefinedGroupId
+        // EndpointStore(fabric_index, endpoint_id).Save(endpoint_info);
+
+        // Remove endpoint info
+
+        if (kInvalidEndpointId == prev_endpoint)
+        {
+            // First endpoint, update fabric info
+            fabric_info.first_endpoint = endpoint_info.next;
+            fabric_store.Save(fabric_info);
+        }
+        else
+        {
+            // Mid endpoint, update previous endpoint's info
+            prev_endpoint_info.next = endpoint_info.next;
+            EndpointStore(fabric_index, prev_endpoint).Save(prev_endpoint_info);
+        }
+
+        return EndpointStore(fabric_index, target_endpoint_id).Delete();
+    }
+
```

Comment:
I was hoping for a endpoint Getter method to enable Attributes Write and Command invocation. Something along the line of : 
`CHIP_ERROR getEndpointId(FabricId fabricId, GroupId gid, EndpointId * endpointId);` 

could you add something like this ?

Reply:
What would that getter do, exactly?  A group can correspond to multiple endpoints.

Reply:
I did a major clean-up in order to use the PersistentStorageDelegate interface. The new line now reads:
```
return endpoint.Delete(mStorage);
```